Added-Real-Time-Face-Detection-Function

meenakshi3151 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

In the main.py file, I have added the code in the real-time-detection which  captures the video frame and calls for the recognize face function which will recognize the face using trained data. If the person is not in the trained data then he/she will be annotated as Unknown .

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

Fixes #1 

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Motivation and Context

This change will make the system more dynamic by capturing the real time video.

## How Has This Been Tested?

This has been tested by showing different images to the camera which are in the trained model and it is able to annotate the images correctly .

## Screenshots (if appropriate):

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have registered myself at [Contrihub website](https://sac.mnnit.ac.in/contrihub).
- [x] My code follows the code style of this project.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have added tests to cover my changes.
- [x ] All new and existing tests passed.


**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

meenakshi3151 committed:
real-time-detection

meenakshi3151 committed:
real-time-face-detection

meenakshi3151 committed:
Update main.py

sanskaromar commented:
What's with this PR name? 
Additionally, Get issue reassigned to you from Contrihub and make a PR request there.

Update this `Fixes # (issue)` with the issue number this PR fixes.

sanskaromar commented on main.py:
```diff
@@ -62,6 +121,7 @@ def main():
 
             if sub_choice == '1':
                 real_time_detection(security_system)
+              
             elif sub_choice == '2':
```
Why this newline here?

sanskaromar commented on main.py:
```diff
@@ -8,20 +8,79 @@ def train_new_model():
-    # Capture video from a camera (you can adjust the video source)
     
+    capture_vid = cv2.VideoCapture(0)  
+    # Now perform face recognition on the frame
+    #Now while the condition is true , we will capture the image
```
Inconsistent spacing in comments.

sanskaromar commented on main.py:
```diff
@@ -8,20 +8,79 @@ def train_new_model():
+    while True:
+        ret, frame = capture_vid.read()
         # Perform face recognition on the frame
+        recognized_faces = recognize_face_dummy(frame)
 
+       
+        for (x, y, w, h) in recognized_faces:
+            cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)  
+        cv2.imshow('Real-time Face Recognition', frame)
+        if cv2.waitKey(1) & 0xFF == ord("q"):
+          break
```
There was no need to implement all this. This is already implemented in `app/security_system.py`. Use face recognizer from there.

sanskaromar commented on main.py:
```diff
@@ -8,20 +8,79 @@ def train_new_model():
+        cv2.imshow('Real-time Face Recognition', frame)
+        if cv2.waitKey(1) & 0xFF == ord("q"):
+          break
         # Press 'q' to quit
+       
+
```
Where is this functionality?

meenakshi3151 replied:
Sir line 75 and 76 contains this functionality

sanskaromar replied:
Okay. Then write this comment there.

sanskaromar commented on main.py:
```diff
@@ -8,20 +8,79 @@ def train_new_model():
+#recognizing using LBPG algo
+face_recognizer = cv2.face.LBPHFaceRecognizer_create()
+#extracting informations
+model_file = 'app/trained_models/trained_face_model_LBPH.xml'  # Update with the appropriate model file
+
+try:
+    face_recognizer.read(model_file)
+except cv2.error as e:
+    print(f"We are  unable to open model file '{model_file}': {e}")
+    exit()
+label_to_name = {}
+label_to_name_file = 'app/label_to_name.csv'  # you should update with the appropriate label-to-name file
+try:
+    with open(label_to_name_file) as f:
+        for line in f:
+            label, name = line.strip().split(',')
+            label_to_name[int(label)] = name
+except IOError:
+    print(f"Unable to open label-to-name file '{label_to_name_file}'")
+    exit()
+    
+def recognize_face_dummy(frame):
+ 
+    # recognized_faces = [(50, 50, 100, 100), (200, 200, 80, 80)]  
+    # return recognized_faces
+    gray_image = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+    faces = face_classifier.detectMultiScale(gray_image, 1.1, 5, minSize=(40, 40))
+    for (x, y, w, h) in faces:
+         label, confidence = face_recognizer.predict(gray_image[y:y+h, x:x+w])
+    
+         if confidence < 50:
+            recognized_label = label_to_name[label]
+         else:
+             recognized_label = "Unknown"
+         cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 4)
+
+         cv2.putText(frame, recognized_label, (x, y-10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (255, 100, 0), 2)
+    return faces
+
+
```
You did all this extra work when there was no need for this. 😔

**It's ALWAYS a great idea to read README and documentation if they are given.**

meenakshi3151 replied:
Ok sir

sanskaromar commented on main.py:
```diff
@@ -8,20 +8,79 @@ def train_new_model():
     selected_algorithm = input("Select the algorithm for training (LBPH/Eigen/Fisher): ").strip()
     train_face_recognition_model(selected_algorithm)
 
+face_classifier = cv2.CascadeClassifier(
+    cv2.data.haarcascades + "haarcascade_frontalface_default.xml"
+)
+
```
:(

sanskaromar reviewed (changes requested):
Please make required changes.

meenakshi3151 committed:
Update main.py

meenakshi3151 committed:
Update main.py

meenakshi3151 committed:
Update main.py

I have added the recognize_face function from the security_system.py by creating the instance of object.

meenakshi3151 committed:
Update security_system.py

meenakshi3151 committed:
Update main.py

meenakshi3151 committed:
Update security_system.py

sanskaromar commented:
Please resolve conflicts in `security_system.py`.
